File: packages/agent-prism/src/tracer.ts
import { CostCalculator } from './cost.js';
import { TraceContext } from './context.js';
import { SqliteStorageAdapter } from './storage/sqlite.js';
import { PostgresStorageAdapter } from './storage/postgres.js';
import type {
  AgentRun,
  CostModel,
  EndRunOptions,
  LogParser,
  ModelCall,
  OnErrorMode,
  RecordModelCallInput,
  RecordToolCallInput,
  Session,
  StartRunOptions,
  StorageAdapter,
  TokenUsage,
  ToolCall,
  ToolCallOptions,
  TraceStatus,
  TracerOptions,
  WrapOptions
} from './types.js';
import { createId } from './utils/id.js';
import { normalizeTokens } from './utils/json.js';

type AnyFunction = (...args: any[]) => any;

export class AgentRunHandle {
  constructor(private readonly tracer: Tracer, readonly id: string, readonly sessionId: string) {}

  toolCall<T>(toolName: string, input: unknown, callback: () => T, options: Omit<ToolCallOptions, 'runId' | 'input'> = {}): T {
    return this.tracer.toolCall(toolName, { ...options, runId: this.id, input }, callback);
  }

  recordToolCall(input: Omit<RecordToolCallInput, 'runId'>): void {
    this.tracer.recordToolCall({ ...input, runId: this.id });
  }

  recordModelCall(input: Omit<RecordModelCallInput, 'runId'>): void {
    this.tracer.recordModelCall({ ...input, runId: this.id });
  }

  startChild(agentName: string, options: Omit<StartRunOptions, 'sessionId' | 'parentRunId' | 'triggeredBy'> = {}): AgentRunHandle {
    return this.tracer.startRun(agentName, {
      ...options,
      sessionId: this.sessionId,
      parentRunId: this.id,
      triggeredBy: 'agent'
    });
  }

  runInContext<T>(callback: () => T): T {
    return this.tracer.withActiveRun(this.id, this.sessionId, callback);
  }

  end(options: EndRunOptions = {}): void {
    this.tracer.endRun(this, options);
  }

  fail(error: unknown): void {
    this.tracer.failRun(this, error);
  }
}

export class Tracer {
  readonly storage: StorageAdapter;
  readonly context = new TraceContext();
  readonly cost: CostCalculator;
  readonly parsers: LogParser[];
  private readonly onError: OnErrorMode;
  private readonly serviceName: string;
  private readonly maxDbSizeBytes?: number;

  constructor(options: TracerOptions = {}) {
    this.storage = resolveStorage(options);
    this.cost = new CostCalculator(options.models);
    this.parsers = (options.parsers ?? []).filter((parser): parser is LogParser => typeof parser === 'object');
    this.onError = options.onError ?? 'warn';
    this.serviceName = options.serviceName ?? 'agent-prism';
    this.maxDbSizeBytes = options.maxDbSizeBytes;
    this.safeStorage('initialize storage', () => this.storage.init());

    if (typeof process !== 'undefined' && process.once) {
      process.once('beforeExit', () => {
        this.flush();
      });
    }
  }

  startRun(agentName: string, options: StartRunOptions = {}): AgentRunHandle {
    const active = this.context.get();
    const now = new Date();
    const sessionId = options.sessionId ?? active?.sessionId ?? createId('ses');
    const parentRunId = options.parentRunId ?? active?.runId;
    const run: AgentRun = {
      id: createId('run'),
      sessionId,
      agentName,
      parentRunId,
      triggeredBy: options.triggeredBy ?? (parentRunId ? 'agent' : 'human'),
      input: options.input,
      status: 'running',
      startedAt: now,
      toolCalls: [],
      modelCalls: [],
      agentCalls: [],
      tokens: normalizeTokens(),
      costUSD: 0,
      metadata: { serviceName: this.serviceName, ...(options.metadata ?? {}) }
    };
    const session: Session = {
      id: sessionId,
      name: options.sessionName ?? agentName,
      status: 'running',
      startedAt: now,
      totalTokens: normalizeTokens(),
      totalCostUSD: 0,
      metadata: options.metadata
    };

    this.safeStorage('create session', () => this.storage.createSession(session));
    this.safeStorage('start run', () => this.storage.upsertAgentRun(run));
    return new AgentRunHandle(this, run.id, sessionId);
  }

  endRun(runOrHandle: AgentRunHandle | string, options: EndRunOptions = {}): void {
    const runId = typeof runOrHandle === 'string' ? runOrHandle : runOrHandle.id;
    const stored = this.readRun(runId);
    if (!stored) {
      this.handleTracingError(new Error(`Cannot end missing run ${runId}`), 'end run');
      return;
    }
    const endedAt = new Date();
    const explicitTokens = options.tokens ? normalizeTokens(options.tokens) : stored.tokens;
    const run: AgentRun = {
      ...stored,
      output: options.output,
      status: options.status ?? 'success',
      error: options.error ?? stored.error,
      errorStack: options.errorStack ?? stored.errorStack,
      endedAt,
      latencyMs: endedAt.getTime() - stored.startedAt.getTime(),
      tokens: explicitTokens,
      costUSD: options.costUSD ?? stored.costUSD,
      metadata: { ...(stored.metadata ?? {}), ...(options.metadata ?? {}) }
    };
    this.safeStorage('end run', () => this.storage.upsertAgentRun(run));
    this.safeStorage('refresh session', () => this.storage.refreshSession(run.sessionId));
    this.maybePrune();
  }

  failRun(runOrHandle: AgentRunHandle | string, error: unknown): void {
    this.endRun(runOrHandle, {
      status: 'failed',
      output: undefined,
      error: errorMessage(error),
      errorStack: errorStack(error),
      metadata: {
        failureCapturedAt: new Date().toISOString()
      }
    });
  }

  timeoutRun(runOrHandle: AgentRunHandle | string, error: unknown = 'Timed out'): void {
    const runId = typeof runOrHandle === 'string' ? runOrHandle : runOrHandle.id;
    const stored = this.readRun(runId);
    if (!stored) return;
    this.safeStorage('record run timeout', () => this.storage.upsertAgentRun({
      ...stored,
      status: 'timeout',
      error: errorMessage(error),
      errorStack: errorStack(error),
      endedAt: new Date(),
      latencyMs: Date.now() - stored.startedAt.getTime()
    }));
    this.safeStorage('refresh timeout session', () => this.storage.refreshSession(stored.sessionId));
  }

  wrap<T extends AnyFunction>(agentName: string, callback: T, options: WrapOptions = {}): T {
    const tracer = this;
    return function wrappedAgent(this: unknown, ...args: Parameters<T>) {
      const input = typeof options.input === 'function' ? options.input(args) : options.input ?? (args.length === 1 ? args[0] : args);
      const run = tracer.startRun(agentName, { ...options, input });

      const execute = () => callback.apply(this, args);
      const finish = (output: unknown) => {
        tracer.endRun(run, { output });
        return output;
      };
      const fail = (error: unknown) => {
        tracer.failRun(run, error);
        throw error;
      };

      try {
        const result = tracer.withActiveRun(run.id, run.sessionId, execute);
        if (isPromiseLike(result)) {
          return result.then(finish, fail);
        }
        return finish(result);
      } catch (error) {
        return fail(error);
      }
    } as T;
  }

  toolCall<T>(toolName: string, input: unknown, callback: () => T): T;
  toolCall<T>(toolName: string, options: ToolCallOptions, callback: () => T): T;
  toolCall<T>(toolName: string, inputOrOptions: unknown, callback: () => T): T {
    const options: ToolCallOptions = isToolCallOptions(inputOrOptions) ? inputOrOptions : { input: inputOrOptions };
    const active = this.context.get();
    const implicitRun = options.runId || active ? undefined : this.startRun(`tool:${toolName}`, { input: options.input, triggeredBy: 'human' });
    const runId = options.runId ?? active?.runId ?? implicitRun!.id;
    const sessionId = active?.sessionId ?? implicitRun?.sessionId ?? this.readRun(runId)?.sessionId;
    const startedAt = Date.now();
    const calledAt = new Date();
    const finish = (output: unknown) => {
      this.recordToolCall({
        ...options,
        runId,
        toolName,
        output,
        latencyMs: Date.now() - startedAt,
        status: 'success',
        calledAt
      });
      if (implicitRun) {
        this.endRun(implicitRun, { output });
      }
      return output;
    };
    const fail = (error: unknown) => {
      this.recordToolCall({
        ...options,
        runId,
        toolName,
        latencyMs: Date.now() - startedAt,
        status: 'failed',
        error: errorMessage(error),
        errorStack: errorStack(error),
        calledAt
      });
      if (implicitRun) {
        this.failRun(implicitRun, error);
      }
      throw error;
    };

    try {
      const result = sessionId ? this.withActiveRun(runId, sessionId, callback) : callback();
      if (isPromiseLike(result)) {
        return result.then(finish, fail) as T;
      }
      return finish(result) as T;
    } catch (error) {
      return fail(error) as T;
    }
  }

  recordToolCall(input: RecordToolCallInput): void {
    const active = this.context.get();
    const implicitRun = input.runId || active ? undefined : this.startRun(`tool:${input.toolName}`, { input: input.input, triggeredBy: 'human' });
    const call: ToolCall = {
      id: input.id ?? createId('tool'),
      runId: input.runId ?? active?.runId ?? implicitRun!.id,
      toolName: input.toolName,
      input: input.input,
      output: input.output,
      latencyMs: input.latencyMs ?? 0,
      tokens: normalizeTokens(input.tokens),
      costUSD: input.costUSD ?? 0,
      status: input.status ?? (input.error ? 'failed' : 'success'),
      error: input.error,
      errorStack: input.errorStack,
      calledAt: input.calledAt ?? new Date(),
      metadata: input.metadata
    };
    this.safeStorage('record tool call', () => this.storage.insertToolCall(call));
    if (implicitRun) {
      this.endRun(implicitRun, { output: input.output });
    }
  }

  recordModelCall(input: RecordModelCallInput): void {
    const active = this.context.get();
    const implicitRun = input.runId || active ? undefined : this.startRun(`model:${input.provider}`, { input: input.input, triggeredBy: 'human' });
    const tokens = normalizeTokens(input.tokens);
    const call: ModelCall = {
      id: input.id ?? createId('model'),
      runId: input.runId ?? active?.runId ?? implicitRun!.id,
      provider: input.provider,
      method: input.method,
      model: input.model,
      input: input.input,
      output: input.output,
      latencyMs: input.latencyMs ?? 0,
      tokens,
      costUSD: input.costUSD ?? this.cost.calculate(input.model, tokens),
      status: input.status ?? (input.error ? 'failed' : 'success'),
      error: input.error,
      errorStack: input.errorStack,
      calledAt: input.calledAt ?? new Date(),
      metadata: input.metadata
    };
    this.safeStorage('record model call', () => this.storage.insertModelCall(call));
    if (implicitRun) {
      this.endRun(implicitRun, { output: input.output });
    }
  }

  withActiveRun<T>(runId: string, sessionId: string, callback: () => T): T {
    return this.context.run({ runId, sessionId }, callback);
  }

  currentRunId(): string | undefined {
    return this.context.get()?.runId;
  }

  flush(): void {
    if (this.maxDbSizeBytes) {
      this.maybePrune();
    }
  }

  shutdown(): void {
    this.flush();
    this.safeStorage('close storage', () => this.storage.close());
  }

  private readRun(runId: string): AgentRun | undefined {
    try {
      return this.storage.getRun(runId) as AgentRun | undefined;
    } catch (error) {
      this.handleTracingError(error, 'read run');
      return undefined;
    }
  }

  private safeStorage(label: string, callback: () => unknown): void {
    try {
      const result = callback();
      if (isPromiseLike(result)) {
        result.catch((error: unknown) => this.handleTracingError(error, label));
      }
    } catch (error) {
      this.handleTracingError(error, label);
    }
  }

  private maybePrune(): void {
    const maxDbSizeBytes = this.maxDbSizeBytes;
    if (!maxDbSizeBytes || !this.storage.prune) {
      return;
    }
    this.safeStorage('prune database', () => this.storage.prune!(maxDbSizeBytes));
  }

  private handleTracingError(error: unknown, label: string): void {
    if (this.onError === 'silent') {
      return;
    }
    const wrapped = error instanceof Error ? error : new Error(String(error));
    wrapped.message = `Agent Prism ${label} failed: ${wrapped.message}`;
    if (this.onError === 'throw') {
      throw wrapped;
    }
    console.warn(wrapped.message);
  }
}

export function createTracer(options: TracerOptions = {}): Tracer {
  return new Tracer(options);
}

function resolveStorage(options: TracerOptions): StorageAdapter {
  const configured = options.storage;
  if (configured && typeof configured === 'object' && 'init' in configured) {
    return configured;
  }
  if (configured && typeof configured === 'object' && configured.type === 'postgres') {
    return new PostgresStorageAdapter(configured);
  }
  if (configured && typeof configured === 'object' && configured.type === 'sqlite') {
    return new SqliteStorageAdapter({ dbPath: configured.dbPath ?? options.dbPath, maxDbSizeBytes: configured.maxDbSizeBytes });
  }
  return new SqliteStorageAdapter({ dbPath: options.dbPath, maxDbSizeBytes: options.maxDbSizeBytes });
}

function isPromiseLike<T = unknown>(value: unknown): value is Promise<T> {
  return Boolean(value && typeof (value as Promise<T>).then === 'function');
}

function isToolCallOptions(value: unknown): value is ToolCallOptions {
  if (!value || typeof value !== 'object') {
    return false;
  }
  return ['runId', 'input', 'tokens', 'costUSD', 'metadata'].some((key) => key in value);
}

function errorMessage(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }
  return String(error);
}

function errorStack(error: unknown): string | undefined {
  return error instanceof Error ? error.stack : undefined;
}

export type { CostModel, TokenUsage, TraceStatus };

File: packages/agent-prism-py/agent_prism/tracer.py
from __future__ import annotations

import contextvars
import time
import traceback
from datetime import datetime
from functools import wraps
from typing import Any, Callable, TypeVar
from uuid import uuid4

from .storage import SqliteStorage
from .types import AgentRun, TokenUsage

T = TypeVar("T")
_active_run: contextvars.ContextVar[tuple[str, str] | None] = contextvars.ContextVar("agent_prism_active_run", default=None)


def create_id(prefix: str) -> str:
    return f"{prefix}_{uuid4().hex[:10]}"


class RunHandle:
    def __init__(self, tracer: "Tracer", run_id: str, session_id: str) -> None:
        self.tracer = tracer
        self.id = run_id
        self.session_id = session_id

    def tool_call(self, name: str, input_value: Any, callback: Callable[[], T]) -> T:
        return self.tracer.tool_call(name, input_value, callback, self.id)

    def end(self, output: Any = None) -> None:
        self.tracer.end_run(self, output=output)


class Tracer:
    def __init__(self, db_path: str = "./agent-prism.db", on_error: str = "warn") -> None:
        self.storage = SqliteStorage(db_path)
        self.on_error = on_error

    def start_run(self, agent_name: str, input: Any = None, session_id: str | None = None, parent_run_id: str | None = None, triggered_by: str | None = None) -> RunHandle:
        active = _active_run.get()
        active_run_id, active_session_id = active if active else (None, None)
        next_session_id = session_id or active_session_id or create_id("ses")
        next_parent_id = parent_run_id or active_run_id
        now = datetime.utcnow()
        self.storage.create_session(next_session_id, agent_name, now)
        run = AgentRun(
            id=create_id("run"),
            session_id=next_session_id,
            agent_name=agent_name,
            parent_run_id=next_parent_id,
            triggered_by=triggered_by or ("agent" if next_parent_id else "human"),
            input=input,
            started_at=now,
        )
        self.storage.upsert_run(run)
        return RunHandle(self, run.id, next_session_id)

    def end_run(self, run: RunHandle, output: Any = None, status: str = "success", tokens: TokenUsage | None = None, cost_usd: float = 0.0) -> None:
        row = self.storage.connection.execute("SELECT * FROM agent_runs WHERE id=?", (run.id,)).fetchone()
        if not row:
            return
        started_at = datetime.fromisoformat(str(row["started_at"]).replace("Z", ""))
        ended_at = datetime.utcnow()
        agent_run = AgentRun(
            id=run.id,
            session_id=run.session_id,
            agent_name=row["agent_name"],
            parent_run_id=row["parent_run_id"],
            triggered_by=row["triggered_by"],
            input=None,
            output=output,
            status=status,  # type: ignore[arg-type]
            started_at=started_at,
            ended_at=ended_at,
            latency_ms=int((ended_at - started_at).total_seconds() * 1000),
            tokens=tokens or TokenUsage(),
            cost_usd=cost_usd,
        )
        self.storage.upsert_run(agent_run)

    def trace(self, agent_name: str) -> Callable[[Callable[..., T]], Callable[..., T]]:
        def decorator(function: Callable[..., T]) -> Callable[..., T]:
            @wraps(function)
            def wrapper(*args: Any, **kwargs: Any) -> T:
                run = self.start_run(agent_name, input={"args": args, "kwargs": kwargs})
                token = _active_run.set((run.id, run.session_id))
                try:
                    result = function(*args, **kwargs)
                    self.end_run(run, output=result)
                    return result
                except Exception as error:
                    self._fail_run(run, error)
                    raise
                finally:
                    _active_run.reset(token)
            return wrapper
        return decorator

    def tool_call(self, name: str, input_value: Any, callback: Callable[[], T], run_id: str | None = None) -> T:
        active = _active_run.get()
        target_run_id = run_id or (active[0] if active else None)
        if not target_run_id:
            implicit = self.start_run(f"tool:{name}", input=input_value)
            target_run_id = implicit.id
        started = time.perf_counter()
        try:
            output = callback()
            self.storage.insert_tool_call(target_run_id, create_id("tool"), name, input_value, output, int((time.perf_counter() - started) * 1000))
            return output
        except Exception as error:
            self.storage.insert_tool_call(target_run_id, create_id("tool"), name, input_value, None, int((time.perf_counter() - started) * 1000), "failed", str(error))
            raise

    def _fail_run(self, run: RunHandle, error: Exception) -> None:
        self.end_run(run, status="failed")
        self.storage.connection.execute(
            "UPDATE agent_runs SET error=?, error_stack=?, status='failed' WHERE id=?",
            (str(error), "".join(traceback.format_exception(error)), run.id),
        )
        self.storage.connection.commit()

    def close(self) -> None:
        self.storage.close()


def create_tracer(db_path: str = "./agent-prism.db") -> Tracer:
    return Tracer(db_path=db_path)